src/cargo/sources/config.rs

```diff
@@ -0,0 +1,149 @@
+        let mut base = SourceConfigMap {
+            cfgs: HashMap::new(),
+            id2name: HashMap::new(),
+            config: config,
+        };
+        base.add("crates-io", SourceConfig {
+            id: try!(SourceId::crates_io(config)),
+            replace_with: None,
+        });
+        Ok(base)
+    }
+
+    pub fn load(&self, id: &SourceId) -> CargoResult<Box<Source + 'cfg>> {
+        debug!("loading: {}", id);
+        let mut name = match self.id2name.get(id) {
+            Some(name) => name,
+            None => return Ok(id.load(self.config)),
+        };
+        let mut path = Path::new("/");
+        let orig_name = name;
+        loop {
+            let cfg = match self.cfgs.get(name) {
+                Some(cfg) => cfg,
+                None => bail!("could not find a configured source with the \
+                               name `{}` when attempting to lookup `{}` \
+                               (configuration in `{}`)",
+                              name, orig_name, path.display()),
+            };
+            match cfg.replace_with {
+                Some((ref s, ref p)) => {
+                    name = s;
+                    path = p;
+                }
+                None if *id == cfg.id => return Ok(id.load(self.config)),
+                None => {
+                    let new_id = cfg.id.with_precise(id.precise()
+                                                       .map(|s| s.to_string()));
+                    let src = new_id.load(self.config);
+                    return Ok(Box::new(ReplacedSource::new(id, &new_id, src)))
+                }
+            }
+            debug!("following pointer to {}", name);
+            if name == orig_name {
+                bail!("detected a cycle of `replace-with` sources, the source \
+                       `{}` is eventually replaced with itself \
+                       (configuration in `{}`)", name, path.display())
+            }
+        }
+    }
+
+    fn add(&mut self, name: &str, cfg: SourceConfig) {
+        self.id2name.insert(cfg.id.clone(), name.to_string());
+        self.cfgs.insert(name.to_string(), cfg);
+    }
+
+    fn add_config(&mut self, name: &str, cfg: &ConfigValue) -> CargoResult<()> {
+        let (table, _path) = try!(cfg.table(&format!("source.{}", name)));
+        let mut srcs = Vec::new();
+        if let Some(val) = table.get("registry") {
+            let url = try!(url(val, &format!("source.{}.registry", name)));
+            srcs.push(SourceId::for_registry(&url));
+        }
+
+        let mut srcs = srcs.into_iter();
+        let src = try!(srcs.next().chain_error(|| {
+            human(format!("no source URL specified for `source.{}`, needs \
+                           `registry` defined", name))
+        }));
+        if srcs.next().is_some() {
+            return Err(human(format!("more than one source URL specified for \
+                                      `source.{}`", name)))
+        }
```

Comment:
Why is `srcs` a `Vec` here if exactly one is required? Seems like a lot of extra error handling.